### PR TITLE
Clarify in docs how runtime parameter resolution works

### LINKED ISSUE #3456: Fix the overwriting of nested parameters with `OmegaConfigLoader`
## Introduction

A user point out that overring nested parameters was not documented. It was not a well known feature and it was only possible via CLI in early 0.18 series.

Add an example of how to do this via CLI or should use the `runtime_params` resolver as user pointed out.

> Hi team, a question when using Kedro params at runtime:
> Right now i have a project with the following parameters.yml (example):
> aaa:
>   bbb:
>     aba: "2023-11-01"
>     abb: "2023-11-01"
>     abc: 14
> I was using Kedro 0.18.2, and running kedro command with the following works fine:
> kedro run --pipeline deploy --params="aaa.bbb.aba=2023-12-01, aaa.bbb.abb=2023-12-01"
> Recently we are deploying a new image which got the version bumped to 0.18.14, and now the pipelines throws the following error:
> KeyError: 'abc'
> which implies aaa.bbb.abc in this nested parameter is not presented anymore. May I know is it an expected change? We read the docs but it does not specify nested parameter overwriting rules at runtime (https://docs.kedro.org/en/latest/configuration/parameters.html#how-to-specify-parameters-at-runtime)
> and what would the suggested solution for such case? Thanks in advance :pray:

### PR DESCRIPTION
**Pull request by merelcht**
## Description
Closes #3456

## Development notes
The problem is that when runtime parameters are provided, they get merged with both the config loaded from the base environment as well as the run environment. So they are duplicated and because of the destructive merge happening by default this can end up removing values present in base config and not in the runtime config.

Solutions tried:
1. Only merge `runtime_params` into base config and not in all loaded configs: this meets the requirements for point 1 and 3 above, but not 2. 
2. Move the merging of `runtime_params` to after both `base` and `default_run_env` config have been loaded. This meets requirements for 1 and 2, but not 3.

### **Proposed "solution"**

not change anything. All the behaviours can be met with the current implementation in place, but to meet the expectation described in #3456  the user has to specify **soft merging** for parameters to ensure the `default_run_env` doesn't destructively overwrite (nested) parameters in `base` config. 

## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

**Review by noklam — COMMENTED** [review]
I found it difficult to understand conceptually that we only include runtime_params for base_env but not the others. I think this may be buggy:

I try this with an example:

```
# base
model_options:
  test_size: 0.2
  random_state: 3
  features:
    - engines


```

```
# local
model_options:
  test_size: 0.2
  random_state: 3
  features:
    - engines
nok: 1
```

```python
%reload_kedro --params nok=3
config = context.config_loader
config["parameters"]
```

The expexted parameters should give me `nok=3`, but instead I get `nok=1`. If I remove `nok:1` from `local` then I get `nok=3` again.

**Comment by merelcht** [comment]
@noklam you're right. Let me dig some more!

**Comment by ElenaKhaustova** [comment]
What if we just merge `base` and `default_run_env` first and then merge `runtime_params` once? From what I got, the problem is that `base` and `default_run_env` have different priorities when merging, and merging `runtime_params` to each leads to losing some parameters from `runtime_params`. So, if we merge `runtime_params` on top user can always define how to do the last (`runtime_params`) merge, whether they override the result of the previous merge or not.

**Review by DimedS — APPROVED** [review]
Thanks for digging into that so thoroughly, @merelcht! From what I understand, the desired behavior is achievable, but we need to clearly document it for the users, right?

**Comment by merelcht** [comment]
> What if we just merge `base` and `default_run_env` first and then merge `runtime_params` once? From what I got, the problem is that `base` and `default_run_env` have different priorities when merging, and merging `runtime_params` to each leads to losing some parameters from `runtime_params`. So, if we merge `runtime_params` on top user can always define how to do the last (`runtime_params`) merge, whether they override the result of the previous merge or not.

I thought the same and tried this, but then this part doesn't work anymore: 
> 3. Any interpolated values are overwritten by runtime_params when provided (see test: test_runtime_params_override_interpolated_value)

**Comment by ElenaKhaustova** [comment]
> > What if we just merge `base` and `default_run_env` first and then merge `runtime_params` once? From what I got, the problem is that `base` and `default_run_env` have different priorities when merging, and merging `runtime_params` to each leads to losing some parameters from `runtime_params`. So, if we merge `runtime_params` on top user can always define how to do the last (`runtime_params`) merge, whether they override the result of the previous merge or not.
> 
> @ElenaKhaustova I thought the same and tried this, but then this part doesn't work anymore:
> 
> > 3. Any interpolated values are overwritten by runtime_params when provided (see test: test_runtime_params_override_interpolated_value)

Then, it looks like there's no easy way to fix that without refactoring the logic. So, just explaining how to achieve the desired behaviour looks good, though it's not easy cause the examples are quite niche and require an understanding of how resolution works from the user side. We can keep an eye on this case in case we decide to refactor this part to avoid corner cases cause it looks quite complex now.

**Review by DimedS — APPROVED** [review]
Thank you, @merelcht! I'm not sure, but perhaps including an example to the docs similar to the one you provided in the tests would help to better illustrate the concept.

File: docs/source/configuration/advanced_configuration.md [hunk 1]
```diff
@@ -322,7 +322,8 @@ Note that you can only use the resolver in `credentials.yml` and not in catalog
 ```
 
 ### How to change the merge strategy used by `OmegaConfigLoader`
-By default, `OmegaConfigLoader` merges configuration [in different environments](configuration_basics.md#configuration-environments) in a destructive way. This means that whatever configuration resides in your overriding environment (`local` by default) takes precedence when the same top-level key is present in the base and overriding environment. Any configuration for that key **besides that given in the overriding environment** is discarded.
+By default, `OmegaConfigLoader` merges configuration [in different environments](configuration_basics.md#configuration-environments) as well as runtime parameters in a destructive way. This means that whatever configuration resides in your overriding environment (`local` by default) takes precedence when the same top-level key is present in the base and overriding environment. Any configuration for that key **besides that given in the overriding environment** is discarded.
+The same behaviour applies to runtime parameters overriding any configuration in the `base` environment.
```
[the remark this answers is not in the document]
Comment (astrojuanlu) [reply docs/source/configuration/advanced_configuration.md:326]: Oh I think that's explained in the example below 👍🏼

**Review by astrojuanlu — APPROVED** [review]
💯

**Comment by DimedS** [comment]
> > Thank you, @merelcht! I'm not sure, but perhaps including an example to the docs similar to the one you provided in the tests would help to better illustrate the concept.
> 
> Done! Let me know if this is what you expected.

super, thank you!
